File: sdks/zig/examples/interned-echo.zig
const sf = @import("shopify_function");

var string_buf: [65536]u8 = undefined;

var interned_foo: sf.InternedStringId = undefined;
var interned_bar: sf.InternedStringId = undefined;

fn strEql(a: []const u8, b: []const u8) bool {
    if (a.len != b.len) return false;
    for (a, b) |ac, bc| {
        if (ac != bc) return false;
    }
    return true;
}

fn echoValue(val: sf.Value) void {
    switch (val.getTag()) {
        .null => sf.outputNull() catch return,
        .boolean => sf.outputBool(val.asBool().?) catch return,
        .number => {
            const num = val.asNumber().?;
            const truncated = @trunc(num);
            if (truncated == num and num >= -2147483648.0 and num <= 2147483647.0) {
                sf.outputI32(@intFromFloat(num)) catch return;
            } else {
                sf.outputF64(num) catch return;
            }
        },
        .string => {
            const len = val.stringLen();
            const buf = string_buf[0..len];
            val.readString(buf);
            sf.outputString(buf) catch return;
        },
        .object => {
            const len = val.objLen().?;
            sf.outputObject(len) catch return;
            for (0..len) |i| {
                const key = val.getObjKeyAtIndex(@intCast(i));
                const key_len = key.stringLen();
                const key_buf = string_buf[0..key_len];
                key.readString(key_buf);

                if (strEql(key_buf, "foo")) {
                    // Use interned string for key and interned obj prop for value
                    sf.outputInternedString(interned_foo) catch return;
                    const child = val.getInternedObjProp(interned_foo);
                    echoValue(child);
                } else if (strEql(key_buf, "bar")) {
                    sf.outputInternedString(interned_bar) catch return;
                    const child = val.getInternedObjProp(interned_bar);
                    echoValue(child);
                } else {
                    sf.outputString(key_buf) catch return;
                    const child = val.getAtIndex(@intCast(i));
                    echoValue(child);
                }
            }
            sf.outputFinishObject() catch return;
        },
        .array => {
            const len = val.arrayLen().?;
            sf.outputArray(len) catch return;
            for (0..len) |i| {
                const child = val.getAtIndex(@intCast(i));
                echoValue(child);
            }
            sf.outputFinishArray() catch return;
        },
        .err => sf.outputNull() catch return,
    }
}

export fn _start() void {
    // Intern strings at startup
    interned_foo = sf.internString("foo");
    interned_bar = sf.internString("bar");

    // Log to exercise log API
    sf.log("interned-echo");

    const input = sf.inputGet();
    echoValue(input);
}

File: sdks/c/examples/interned-echo.c
#include "shopify_function_value.h"

// Interned string IDs (populated at startup)
static InternedStringId interned_foo;
static InternedStringId interned_bar;

static const uint8_t STR_FOO[] = "foo";
static const uint8_t STR_BAR[] = "bar";
static const uint8_t LOG_MSG[] = "interned-echo";

// Forward declaration
static void echo_value(Val val);

static void echo_value(Val val) {
    SFValueTag tag = sf_value_get_tag(val);

    switch (tag) {
        case SF_TAG_NULL:
            shopify_function_output_new_null();
            break;

        case SF_TAG_BOOL:
            shopify_function_output_new_bool((uint32_t)sf_value_as_bool(val));
            break;

        case SF_TAG_NUMBER: {
            double num = sf_value_as_number(val);
            double truncated = num >= 0 ? (double)(int64_t)num : (double)(int64_t)num;
            if (truncated == num && num >= -2147483648.0 && num <= 2147483647.0) {
                shopify_function_output_new_i32((int32_t)num);
            } else {
                shopify_function_output_new_f64(num);
            }
            break;
        }

        case SF_TAG_STRING: {
            size_t len = sf_string_len(val);
            uint8_t* buf = sf_bump_alloc(len);
            if (buf) {
                sf_read_string(val, buf, len);
                shopify_function_output_new_utf8_str(buf, len);
            }
            break;
        }

        case SF_TAG_OBJECT: {
            size_t len = sf_object_len(val);
            shopify_function_output_new_object(len);
            for (size_t i = 0; i < len; i++) {
                // Read key to decide if it's internable
                Val key_val = shopify_function_input_get_obj_key_at_index(val, i);
                size_t key_len = sf_string_len(key_val);
                uint8_t* key_buf = sf_bump_alloc(key_len);
                if (!key_buf) break;
                sf_read_string(key_val, key_buf, key_len);

                // Check if key matches "foo" or "bar" for interned output
                int is_foo = (key_len == 3 && key_buf[0] == 'f' && key_buf[1] == 'o' && key_buf[2] == 'o');
                int is_bar = (key_len == 3 && key_buf[0] == 'b' && key_buf[1] == 'a' && key_buf[2] == 'r');

                if (is_foo) {
                    // Write key using interned string
                    shopify_function_output_new_interned_utf8_str(interned_foo);
                    // Read value using interned obj prop
                    Val child = shopify_function_input_get_interned_obj_prop(val, interned_foo);
                    echo_value(child);
                } else if (is_bar) {
                    shopify_function_output_new_interned_utf8_str(interned_bar);
                    Val child = shopify_function_input_get_interned_obj_prop(val, interned_bar);
                    echo_value(child);
                } else {
                    // Regular string key
                    shopify_function_output_new_utf8_str(key_buf, key_len);
                    Val child = shopify_function_input_get_at_index(val, i);
                    echo_value(child);
                }
            }
            shopify_function_output_finish_object();
            break;
        }

        case SF_TAG_ARRAY: {
            size_t len = sf_array_len(val);
            shopify_function_output_new_array(len);
            for (size_t i = 0; i < len; i++) {
                Val child = shopify_function_input_get_at_index(val, i);
                echo_value(child);
            }
            shopify_function_output_finish_array();
            break;
        }

        default:
            shopify_function_output_new_null();
            break;
    }
}

void _start(void) {
    sf_bump_reset();

    // Intern strings at startup
    interned_foo = shopify_function_intern_utf8_str(STR_FOO, sizeof(STR_FOO) - 1);
    interned_bar = shopify_function_intern_utf8_str(STR_BAR, sizeof(STR_BAR) - 1);

    // Log to exercise log API
    shopify_function_log_new_utf8_str(LOG_MSG, sizeof(LOG_MSG) - 1);

    Val input = shopify_function_input_get();
    echo_value(input);
}
